Add support for spread syntax in templates

crisbeto opened the pull request:
Adds support for spread/rest syntax in the following places in the template syntax:
* Object literals, e.g. `{a: 1, ...foo}`.
* Array literals, e.g. `[1, ...foo]`.
* Function calls, e.g. `fn(1, ...foo)`.

For arrays and object literals the spread arguments integrate into the existing pure function infrastructure in order to avoid re-creating the object on each change detection.

Fixes #11850.
Fixes #61800.

Note: I took a couple of commits from #66206 and #66190 to avoid conflicts.

crisbeto committed:
refactor(compiler): handle spread operators in the lexer

Updates the expression lexer to handle spread operators.

crisbeto committed:
refactor(compiler): support spread elements in output AST

Updates the output AST to account for spread elements.

crisbeto committed:
refactor(compiler-cli): account for spread assignments in linker

Updates the linker to account for spread assignments in object literals.

crisbeto committed:
refactor(compiler): add spread elements to expression AST

Updates the expression AST to have support for spread elements inside object literals.

crisbeto committed:
feat(core): support spread expressions in object literals

Adds support for spread expressions inside of object literals. This can be handy when constructing maps for `class` bindings.

crisbeto committed:
refactor(compiler-cli): account for spread elements in translator

Adds the ability to generate spread elements in the linker.

crisbeto committed:
refactor(compiler): add spread elements to the output AST

Adds spread elements to the output AST which can be used in arrays and function calls.

crisbeto committed:
refactor(compiler): add spread elements to expression AST

Updates the expression AST to support spread elements that will be used for arrays and function calls.

crisbeto committed:
feat(core): support spread elements in array literals

Expands the template syntax to support spread elements inside arrays. This can be handy for some bindings.

JeanMeche commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -1302,14 +1325,32 @@ class _ParseAST {
+
     do {
-      positionals.push(this.parsePipe());
+      positionals.push(this.next.isOperator('...') ? this.parseSpreadElement() : this.parsePipe());
```
nit: should we create a custom method `isSpearOperator` ?

JeanMeche reviewed (approved):
Beside the nit, LGTM 🚀

crisbeto committed:
feat(core): support rest arguments in function calls

Updates the template syntax to support rest arguments in function calls. This can be handy for functions with a variable number of arguments.
